connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorMaker.java

```diff
@@ -0,0 +1,243 @@
+        boolean wasShuttingDown = shutdown.getAndSet(true);
+        if (!wasShuttingDown) {
+            log.info("Kafka MirrorMaker stopping");
+            for (Herder herder : herders.values()) {
+                try {
+                    herder.stop();
+                } finally {
+                    stopLatch.countDown();
+                }
+            }
+            log.info("Kafka MirrorMaker stopped.");
+        }
+    }
+
+    public void awaitStop() {
+        try {
+            stopLatch.await();
+        } catch (InterruptedException e) {
+            log.error("Interrupted waiting for MirrorMaker to shutdown");
+        }
+    }
+
+    public void pause(String source, String target) {
+        SourceAndTarget sourceAndTarget = new SourceAndTarget(source, target);
+        checkHerder(sourceAndTarget);
+        Herder herder = herders.get(sourceAndTarget);
+        CONNECTOR_CLASSES.forEach(x -> herder.pauseConnector(x.getSimpleName()));
+    }
+
+    public void resume(String source, String target) {
+        SourceAndTarget sourceAndTarget = new SourceAndTarget(source, target);
+        checkHerder(sourceAndTarget);
+        Herder herder = herders.get(sourceAndTarget);
+        CONNECTOR_CLASSES.forEach(x -> herder.resumeConnector(x.getSimpleName()));
+    }
+
+    private void configureConnector(SourceAndTarget sourceAndTarget, Class connectorClass) {
+        checkHerder(sourceAndTarget);
+        Map<String, String> connectorProps = config.connectorBaseConfig(sourceAndTarget, connectorClass);
+        herders.get(sourceAndTarget)
+                .putConnectorConfig(connectorClass.getSimpleName(), connectorProps, true, (e, x) -> {
+                    if (e instanceof NotLeaderException) {
+                        // No way to determine if the connector is a leader or not beforehand.
+                        log.info("Connector {} is a follower. Using existing configuration.", sourceAndTarget);
+                    } else {
+                        log.info("Connector {} configured.", sourceAndTarget, e);
+                    }
+                });
+    }
+
+    private void checkHerder(SourceAndTarget sourceAndTarget) {
+        if (!herders.containsKey(sourceAndTarget)) {
+            throw new IllegalArgumentException("No herder for " + sourceAndTarget.toString());
+        }
+    }
+
+    private void configureConnectors(SourceAndTarget sourceAndTarget) {
+        CONNECTOR_CLASSES.forEach(x -> configureConnector(sourceAndTarget, x));
+    }
+
+    private void addHerder(SourceAndTarget sourceAndTarget) {
+        log.info("creating herder for " + sourceAndTarget.toString());
+        Map<String, String> workerProps = config.workerConfig(sourceAndTarget);
+        String advertisedUrl = advertisedBaseUrl + "/" + sourceAndTarget.source();
+        String workerId = sourceAndTarget.toString();
+        Plugins plugins = new Plugins(workerProps);
+        plugins.compareAndSwapWithDelegatingLoader();
```

Comment:
Should the returned classloader be restored at the end of the method?

Reply:
I'm not certain, but the other drivers (connect-standalone, connect-distributed) do it this way. I'll just cargo-cult here.